dom-fetch-client/src/main/scala/org/http4s/dom/FetchClient.scala

```diff
@@ -20,30 +20,75 @@ package dom
 import cats.effect.Async
 import cats.effect.Resource
 import cats.syntax.all._
+import cats.effect.syntax.all._
+import java.util.concurrent.TimeoutException
 import org.http4s.client.Client
 import org.scalajs.dom.crypto._
+import org.scalajs.dom.experimental.AbortController
 import org.scalajs.dom.experimental.Fetch
 import org.scalajs.dom.experimental.Headers
 import org.scalajs.dom.experimental.HttpMethod
+import org.scalajs.dom.experimental.ReferrerPolicy
+import org.scalajs.dom.experimental.RequestCache
+import org.scalajs.dom.experimental.RequestCredentials
 import org.scalajs.dom.experimental.RequestInit
+import org.scalajs.dom.experimental.RequestMode
+import org.scalajs.dom.experimental.RequestRedirect
+import org.scalajs.dom.experimental.{Response => FetchResponse}
+import scala.concurrent.duration._
 
 object FetchClient {
 
-  def apply[F[_]](implicit F: Async[F]): Client[F] = Client[F] { req: Request[F] =>
+  private[dom] def makeClient[F[_]](
+      requestTimeout: Duration,
+      cache: Option[RequestCache],
+      credentials: Option[RequestCredentials],
+      integrity: Option[String],
+      keepalive: Option[Boolean],
+      mode: Option[RequestMode],
+      redirect: Option[RequestRedirect],
+      referrer: Option[String],
+      referrerPolicy: Option[ReferrerPolicy]
+  )(implicit F: Async[F]): Client[F] = Client[F] { (req: Request[F]) =>
     Resource.eval(req.body.chunkAll.filter(_.nonEmpty).compile.last).flatMap { body =>
       Resource
         .makeCase {
-          F.fromPromise {
-            F.delay {
-              val init = new RequestInit {}
+          F.delay {
+            val init = new RequestInit {}
 
-              init.method = req.method.name.asInstanceOf[HttpMethod]
-              init.headers = new Headers(toDomHeaders(req.headers))
-              body.foreach { body =>
-                init.body = arrayBuffer2BufferSource(body.toJSArrayBuffer)
-              }
+            init.method = req.method.name.asInstanceOf[HttpMethod]
+            init.headers = new Headers(toDomHeaders(req.headers))
+            body.foreach { body =>
+              init.body = arrayBuffer2BufferSource(body.toJSArrayBuffer)
+            }
+            cache.foreach(init.cache = _)
+            credentials.foreach(init.credentials = _)
+            integrity.foreach(init.integrity = _)
+            keepalive.foreach(init.keepalive = _)
+            mode.foreach(init.mode = _)
+            redirect.foreach(init.redirect = _)
+            referrer.foreach(init.referrer = _)
+            referrerPolicy.foreach(init.referrerPolicy = _)
 
-              Fetch.fetch(req.uri.renderString, init)
+            init
+          } >>= {
+            requestTimeout match {
+              case d: FiniteDuration =>
+                init =>
+                  Resource
+                    .eval(F.delay(new AbortController()))
```

Comment:
This `AbortController` is an important addition, thanks! Instead of `eval`ing it into the resource, maybe it should become part of the resource itself? I.e., in `Resource.makeCase` so that if the `Resource` is cancelled or errored we abort the request.

Reply:
Good catch.